wordpress-mobile/WordPress-iOS
# [Gutenberg] GutenbergSettings refactor

**UPDATE:**

This PR (now) is a refactor of GutenbergSettings and the auto-enable gutenberg Dialog logic, moving all the decision logic out of the `InformativeDialog` extension, making the logic of `auto-enable gutenberg` more generic and accessible.

To test:
- The test cases from https://github.com/wordpress-mobile/WordPress-iOS/pull/12066 should still hold true.
- The case described on https://github.com/wordpress-mobile/WordPress-iOS/pull/12066#pullrequestreview-257903408 should have been fixed with these changes.

You can reset the gutenberg selection calling:
`UserDefaults().set(nil, forKey: GutenbergSettings.Key.enabled)`


----
**Old Description:**

This PR is the WPiOS version of https://github.com/wordpress-mobile/WordPress-Android/pull/10209

WordPressKit side PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/166

On this PR:
- Editor preference is sent to the remote server each time the user switch the pref in App Settings
- The preference is sent to the remote server each time the editor is opened in the app. This was necessary since some users may have already switched to GB, and the new prompt dialog is not shown in these cases. We may want to revise this decision, and send data to the remote server when the dialog is shown only by putting a little bit of logic around it.
- The local app preference is set for all of the wpcom powered sites available in the app


I took the opportunity to move all the decision logic out of the `InformativeDialog` extension, making the logic of `auto-enable gutenberg` more generic and accessible.


To test:
I left a small log for testing purposes.

- Preferably, log in with an alternative account that doesn't have hundreds of posts linked.
- Switch the editor to gutenberg / aztec
  - Check in the report card for that user that the posts are being set accordingly on the backend.
- Open the editor (Gutenberg or Aztec)
  - Check that the backend sets the mobile editor setting accordingly with the app switch selection (not necessarily with the editor that was opened)
  - As an example: If you switch to enable gutenberg, and open non-gutenberg post, it will open Aztec, but since the switch is set to Gutenberg, it should set the backend to Gutenberg anyway.

The test cases from https://github.com/wordpress-mobile/WordPress-iOS/pull/12066 should still hold true, and the case described on https://github.com/wordpress-mobile/WordPress-iOS/pull/12066#pullrequestreview-257903408 should have been fixed with these changes.


Update release notes:
